### PR TITLE
Reduce implementation-dependency of `IStorageServer`

### PR DESCRIPTION
**Pull request by itamarst**
Fixes https://tahoe-lafs.org/trac/tahoe-lafs/ticket/3779

Left in the so-called "pipeline" code, since it's not clear if it's needed or not.

**Update:** Also fixed a pre-existing bug in remote_advise_corrupt_share that was exposed by the changes.

### CODE REVIEW

**Comment by itamarst** [comment]
Hm. Getting the impression that I broke something with the seemingly minor tweaks I did...

**Comment by itamarst** [comment]
Actually—those errors were probably there before, and just were being dropped on the floor, and now they're being logged.

File: src/allmydata/storage/server.py [hunk 1]
```diff
@@ -708,8 +708,10 @@ def remote_advise_corrupt_share(self, share_type, storage_index, shnum,
         now = time_format.iso_utc(sep="T")
         si_s = si_b2a(storage_index)
         # windows can't handle colons in the filename
-        fn = os.path.join(self.corruption_advisory_dir,
-                          "%s--%s-%d" % (now, str(si_s, "utf-8"), shnum)).replace(":","")
+        fn = os.path.join(
+            self.corruption_advisory_dir,
+            ("%s--%s-%d" % (now, str(si_s, "utf-8"), shnum)).replace(":","")
```
Comment (itamarst) [inline src/allmydata/storage/server.py:713]: This is a (easy to miss) bugfix. Previously the replace was on the _whole_ path, thus removing all colons from paths.

File: src/allmydata/immutable/downloader/share.py [hunk 2]
```diff
@@ -475,7 +475,9 @@ def _signal_corruption(self, f, start, offset):
         # there was corruption somewhere in the given range
         reason = "corruption in share[%d-%d): %s" % (start, start+offset,
                                                      str(f.value))
-        self._rref.callRemoteOnly("advise_corrupt_share", reason.encode("utf-8"))
+        self._rref.callRemote(
+            "advise_corrupt_share", reason.encode("utf-8")
+        ).addErrback(log.err)
```
Comment (exarkun) [inline src/allmydata/immutable/downloader/share.py:480]: Passing a descriptive (or at least distinctive) *reason* to `log.err` will make it easier to make sense of any failures logged by this code path.  Can you add that here and to the other new `log.err` errbacks?

File: newsfragments/3779.bugfix [hunk 3]
```diff
@@ -0,0 +1 @@
+Fixed bug where share corruption events were not recorded on Windows.
```
Comment (exarkun) [inline newsfragments/3779.bugfix:1]: Can you make it clear here that this is about behavior for a storage server running on Windows (ie, it doesn't matter where the client is running)?

**Review by exarkun — APPROVED** [review]
Thanks.  Minor comments inline.  Please address and merge.

**Pull request merged**